### PR TITLE
Allows variables starting with underscores

### PR DESCRIPTION
**Pull request by petems**

### CODE REVIEW

File: spec/puppet-spec.coffee [hunk 1]
```diff
@@ -47,3 +47,8 @@ describe "Puppet grammar", ->
       {tokens} = grammar.tokenizeLine("package {$foo:}")
       expect(tokens[0]).toEqual value: 'package', scopes: ['source.puppet', 'meta.definition.resource.puppet', 'storage.type.puppet']
       expect(tokens[2]).toEqual value: '$foo', scopes: ['source.puppet', 'meta.definition.resource.puppet', 'entity.name.section.puppet']
+
+    it "tokenizes variable starting with underscore", ->
+      {tokens} = grammar.tokenizeLine('$foo')
```
Comment (50Wliu) [inline spec/puppet-spec.coffee]: You seem to have forgotten an underscore here... :wink:

Comment (petems) [inline spec/puppet-spec.coffee]: Whoops...

Comment (petems) [inline spec/puppet-spec.coffee]: Fixed and added a test for regular variables too! :+1:

File: spec/puppet-spec.coffee [hunk 2]
```diff
@@ -47,3 +47,13 @@ describe "Puppet grammar", ->
       {tokens} = grammar.tokenizeLine("package {$foo:}")
       expect(tokens[0]).toEqual value: 'package', scopes: ['source.puppet', 'meta.definition.resource.puppet', 'storage.type.puppet']
       expect(tokens[2]).toEqual value: '$foo', scopes: ['source.puppet', 'meta.definition.resource.puppet', 'entity.name.section.puppet']
+
+    it "tokenizes variablee", ->
```
Comment (50Wliu) [inline spec/puppet-spec.coffee]: `it "tokenizes regular variables", ->`

File: spec/puppet-spec.coffee [hunk 3]
```diff
@@ -47,3 +47,13 @@ describe "Puppet grammar", ->
...
+      {tokens} = grammar.tokenizeLine('$foo')
+      expect(tokens[0]).toEqual value: '$', scopes: ['source.puppet', 'variable.other.readwrite.global.puppet', 'punctuation.definition.variable.puppet']
+      expect(tokens[1]).toEqual value: 'foo', scopes: ['source.puppet', 'variable.other.readwrite.global.puppet']
+
+    it "tokenizes variable starting with underscore", ->
```
Comment (50Wliu) [inline spec/puppet-spec.coffee]: This can be grouped under the above test.

File: spec/puppet-spec.coffee [hunk 4]
```diff
@@ -47,3 +47,11 @@ describe "Puppet grammar", ->
       {tokens} = grammar.tokenizeLine("package {$foo:}")
       expect(tokens[0]).toEqual value: 'package', scopes: ['source.puppet', 'meta.definition.resource.puppet', 'storage.type.puppet']
       expect(tokens[2]).toEqual value: '$foo', scopes: ['source.puppet', 'meta.definition.resource.puppet', 'entity.name.section.puppet']
+
+    it "tokenizes regular variables", ->
+      {tokens} = grammar.tokenizeLine('$foo')
+      expect(tokens[0]).toEqual value: '$', scopes: ['source.puppet', 'variable.other.readwrite.global.puppet', 'punctuation.definition.variable.puppet']
+      expect(tokens[1]).toEqual value: 'foo', scopes: ['source.puppet', 'variable.other.readwrite.global.puppet']
+      {tokens} = grammar.tokenizeLine('$_foo')
```
Comment (50Wliu) [inline spec/puppet-spec.coffee]: Tiny little nit: can there be a newline between lines 54 and 55?

Comment (petems) [inline spec/puppet-spec.coffee]: Fixed :+1:

**Comment by 50Wliu** [comment]
One more question: Does Puppet allow variables such as `$_woo__`?

**Comment by petems** [comment]
@50Wliu Yep, that's valid

```puppet
$_woo__='hello'
notice($_woo__)

Notice: Scope(Class[main]): hello
```

File: grammars/puppet.cson [hunk 5]
```diff
@@ -410,7 +410,7 @@
         'captures':
           '1':
             'name': 'punctuation.definition.variable.puppet'
-        'match': '(\\$)([a-zA-Zx7f-xff\\$]|::)([a-zA-Z0-9_x7f-xff\\$]|::)*\\b'
+        'match': '(\\$)([a-zA-Zx7f-xff\\$_]|::)([a-zA-Z0-9_x7f-xff\\$]|::)*\\b'
```
Comment (50Wliu) [inline grammars/puppet.cson]: The latter capture should also include `_` then

**Comment by 50Wliu** [comment]
Can variables not start with uppercase letters?

**Comment by petems** [comment]
In Puppet 4, the official regex as given by the parser: `Error: Illegal variable name, The given name 'Woo' does not conform to the naming rule /^((::)?[a-z]\w*)*((::)?[a-z_]\w*)$/ at /Users/petersouter/echo.pp:1:1`

In older puppet its possible but linters will complain,

**Comment by 50Wliu** [comment]
Wow, that's *pretty* awesome how it gives you the regex right in the error message.  Just waiting on Travis now... :watch: :sleeping:

**Pull request merged**